### PR TITLE
Fix false negative for `used-before-assignment` when an Except intervenes between Try/Finally

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Fixes false negative for `used-before-assignment` when evaluating statements in a finally block, there is a possibly failing statement in the try block, and an except handler intervenes between.

Found while working on #5582. Unhandled case from #5384 -- I had thought of this case, but didn't know how to solve it: a finally that is not the direct descendant of a try handler because we have try/except/finally. (Since we were just incrementally solving false negatives, that wasn't such a problem at the time.) After #5582, I now know how to handle it!

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
Need to test this case -- newly failing on this branch:
```python
try:
    pass
finally:
    try:
        times = 1
    except TypeError:
        pass
    print(times)  # emits. shouldn't. and won't if the outer try goes away.
```

**Comment by DanielNoord** [comment]
@jacobtylerwalls I haven't look at this but in #5582 we discussed doing some of the changes for `used-before-assignment` in `2.13`, get feedback, and then do some changes in `2.14`. Does this fall under the same category of "perhaps wait till 2.14" or should this be reviewed for `2.13`?

**Comment by jacobtylerwalls** [comment]
Hi @DanielNoord thanks for looking into this. I think this is the same category ("no reason to wait; fixes false negatives for assumptions we've already doc'd and merged to master") as #5582, see https://github.com/PyCQA/pylint/pull/5582#issuecomment-1009142789. Happy to hear your views on this.

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -741,6 +733,43 @@ def _uncertain_nodes_in_except_blocks(found_nodes, node, node_statement):
             uncertain_nodes.append(other_node)
         return uncertain_nodes
 
+    @staticmethod
+    def _uncertain_nodes_in_try_blocks_when_evaluating_finally_blocks(
+        found_nodes: List[nodes.NodeNG], node_statement: nodes.Statement
+    ):
```
Comment (DanielNoord) [inline pylint/checkers/variables.py:739]: This is missing return type annotation

**Review by DanielNoord — APPROVED** [review]
@jacobtylerwalls misunderstood! Only one comment 😄

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks good, but the rebase for this is non trivial, and there's a strange conflict in the changelog and whatsnew. Could you rebase yourself @jacobtylerwalls ?

File: tests/functional/u/use/used_before_assignment_issue85.py [hunk 2]
```diff
@@ -7,3 +7,61 @@ def main():
...
+    """
+    try:
+        pass
+    finally:
+        try:
+            times = 1
+        except TypeError:
+            pass
+        # Don't emit: only assume the outer try failed
+        print(times)
```
Comment (Pierre-Sassoulas) [inline tests/functional/u/use/used_before_assignment_issue85.py:67]: I don't understand ``Don't confuse assignments in different finally statements``. Shouldn't we raise or ``times`` be defined somewhere in the other finally statement?

File: tests/functional/c/consider/consider_using_with_open.py [hunk 3]
```diff
@@ -137,7 +137,7 @@ def test_defined_in_try_and_finally(self):
         except FileNotFoundError:
             Path("foo").touch()
         finally:
-            file_handle.open("foo", encoding="utf")  # must not trigger
+            file_handle.open("foo", encoding="utf")  # [used-before-assignment] # must not trigger
```
Comment (Pierre-Sassoulas) [inline tests/functional/c/consider/consider_using_with_open.py:140]: Is it a new false positive or should the comment be removed ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks good already, I have two questions regarding the tests.

File: tests/functional/c/consider/consider_using_with_open.py (hunk 3 above)
Comment (jacobtylerwalls) [reply tests/functional/c/consider/consider_using_with_open.py:140]: I think `used-before-assignment` is correct, but `consider-using-with` is apparently not supposed to be triggered. I can update the comment!

Comment (Pierre-Sassoulas) [reply tests/functional/c/consider/consider_using_with_open.py:140]: Ha, I get it
```suggestion
            # +1: [used-before-assignment] 
            file_handle.open("foo", encoding="utf")  # must not trigger consider-using-with
```

File: tests/functional/u/use/used_before_assignment_issue85.py (hunk 2 above)
Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_issue85.py:67]: The docstring looks like it was lazily copied from above and not really updated. I'll push again and explain in the docstring how I see what is being tested and where to go from here.

Comment (jacobtylerwalls) [reply tests/functional/u/use/used_before_assignment_issue85.py]: In rewriting the tests I discovered I could handle nesting better. So there are more tests now :-)

**Review by Pierre-Sassoulas — APPROVED** [review]
Amazing work, thank you @jacobtylerwalls !

**Pull request merged**